<a href="https://colab.research.google.com/github/AbszG/hotel-price-EDA-and-prediction-model/blob/main/Hotel_Price_Prediction_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Data Preprocessing**

In [ ]:
from google.colab import drive
drive.mount("/content/drive", force_remount=True)

Mounted at /content/drive


In [ ]:
import numpy as np
import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/Dynamic Hotel Booking Price Pred/cleaned_data.csv')
df.head()

Unnamed: 0.1  Unnamed: 0         hotel  is_canceled  lead_time  \
0             0           0  Resort Hotel            0        342   
1             1           1  Resort Hotel            0        737   
2             2           2  Resort Hotel            0          7   
3             3           3  Resort Hotel            0         13   
4             4           4  Resort Hotel            0         14   

   arrival_date_year arrival_date_month  arrival_date_week_number  \
0               2015               July                        27   
1               2015               July                        27   
2               2015               July                        27   
3               2015               July                        27   
4               2015               July                        27   

   arrival_date_day_of_month  stays_in_weekend_nights  ...  agent company  \
0                          1                        0  ...    0.0     0.0   
1                          1                        0  ...    0.0     0.0   
2                          1                        0  ...    0.0     0.0   
3                          1                        0  ...  304.0     0.0   
4                          1                        0  ...  240.0     0.0   

  days_in_waiting_list customer_type   adr  required_car_parking_spaces  \
0                    0     Transient   0.0                            0   
1                    0     Transient   0.0                            0   
2                    0     Transient  75.0                            0   
3                    0     Transient  75.0                            0   
4                    0     Transient  98.0                            0   

   total_of_special_requests  reservation_status reservation_status_date  \
0                          0           Check-Out              2015-07-01   
1                          0           Check-Out              2015-07-01   
2                          0           Check-Out              2015-07-02   
3                          0           Check-Out              2015-07-02   
4                          1           Check-Out              2015-07-03   

  total_guests  
0          2.0  
1          2.0  
2          1.0  
3          1.0  
4          2.0  

[5 rows x 32 columns]

In [ ]:
df.columns

Index(['Unnamed: 0.1', 'Unnamed: 0', 'hotel', 'is_canceled', 'lead_time',
       'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'meal', 'country', 'market_segment',
       'distribution_channel', 'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date', 'total_guests'],
      dtype='object')

In [ ]:
useless_cols = ['Unnamed: 0.1', 'Unnamed: 0' , 'is_canceled' , 'reservation_status' , 'arrival_date_year' , 'arrival_date_month' , 'country']

df.drop(useless_cols, axis = 1, inplace = True)

In [ ]:
df.head()

hotel  lead_time  arrival_date_week_number  \
0  Resort Hotel        342                        27   
1  Resort Hotel        737                        27   
2  Resort Hotel          7                        27   
3  Resort Hotel         13                        27   
4  Resort Hotel         14                        27   

   arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  \
0                          1                        0                     0   
1                          1                        0                     0   
2                          1                        0                     1   
3                          1                        0                     1   
4                          1                        0                     2   

  meal market_segment distribution_channel  is_repeated_guest  ...  \
0   BB         Direct               Direct                  0  ...   
1   BB         Direct               Direct                  0  ...   
2   BB         Direct               Direct                  0  ...   
3   BB      Corporate            Corporate                  0  ...   
4   BB      Online TA                TA/TO                  0  ...   

   deposit_type  agent company days_in_waiting_list  customer_type   adr  \
0    No Deposit    0.0     0.0                    0      Transient   0.0   
1    No Deposit    0.0     0.0                    0      Transient   0.0   
2    No Deposit    0.0     0.0                    0      Transient  75.0   
3    No Deposit  304.0     0.0                    0      Transient  75.0   
4    No Deposit  240.0     0.0                    0      Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  \
0                            0                          0   
1                            0                          0   
2                            0                          0   
3                            0                          0   
4                            0                          1   

   reservation_status_date total_guests  
0               2015-07-01          2.0  
1               2015-07-01          2.0  
2               2015-07-02          1.0  
3               2015-07-02          1.0  
4               2015-07-03          2.0  

[5 rows x 25 columns]

In [ ]:
# creating numerical and categorical dataframes

cat_cols = [col for col in df.columns if df[col].dtype == 'O']
cat_cols

['hotel',
 'meal',
 'market_segment',
 'distribution_channel',
 'reserved_room_type',
 'assigned_room_type',
 'deposit_type',
 'customer_type',
 'reservation_status_date']

In [ ]:
cat_df = df[cat_cols]
cat_df.head()

hotel meal market_segment distribution_channel reserved_room_type  \
0  Resort Hotel   BB         Direct               Direct                  C   
1  Resort Hotel   BB         Direct               Direct                  C   
2  Resort Hotel   BB         Direct               Direct                  A   
3  Resort Hotel   BB      Corporate            Corporate                  A   
4  Resort Hotel   BB      Online TA                TA/TO                  A   

  assigned_room_type deposit_type customer_type reservation_status_date  
0                  C   No Deposit     Transient              2015-07-01  
1                  C   No Deposit     Transient              2015-07-01  
2                  C   No Deposit     Transient              2015-07-02  
3                  A   No Deposit     Transient              2015-07-02  
4                  A   No Deposit     Transient              2015-07-03

In [ ]:
cat_df['reservation_status_date'] = pd.to_datetime(cat_df['reservation_status_date'])

cat_df['year'] = cat_df['reservation_status_date'].dt.year
cat_df['month'] = cat_df['reservation_status_date'].dt.month
cat_df['day'] = cat_df['reservation_status_date'].dt.day

In [ ]:
cat_df.drop(['reservation_status_date'] , axis = 1, inplace = True)

In [ ]:
cat_df.head()

hotel meal market_segment distribution_channel reserved_room_type  \
0  Resort Hotel   BB         Direct               Direct                  C   
1  Resort Hotel   BB         Direct               Direct                  C   
2  Resort Hotel   BB         Direct               Direct                  A   
3  Resort Hotel   BB      Corporate            Corporate                  A   
4  Resort Hotel   BB      Online TA                TA/TO                  A   

  assigned_room_type deposit_type customer_type  year  month  day  
0                  C   No Deposit     Transient  2015      7    1  
1                  C   No Deposit     Transient  2015      7    1  
2                  C   No Deposit     Transient  2015      7    2  
3                  A   No Deposit     Transient  2015      7    2  
4                  A   No Deposit     Transient  2015      7    3

In [ ]:
for col in cat_df.columns:
    print(f"{col}: \n{cat_df[col].unique()}\n")

hotel: 
['Resort Hotel' 'City Hotel']

meal: 
['BB' 'FB' 'HB' 'SC' 'Undefined']

market_segment: 
['Direct' 'Corporate' 'Online TA' 'Offline TA/TO' 'Complementary' 'Groups'
 'Undefined' 'Aviation']

distribution_channel: 
['Direct' 'Corporate' 'TA/TO' 'Undefined' 'GDS']

reserved_room_type: 
['C' 'A' 'D' 'E' 'G' 'F' 'H' 'L' 'P' 'B']

assigned_room_type: 
['C' 'A' 'D' 'E' 'G' 'F' 'I' 'B' 'H' 'P' 'L' 'K']

deposit_type: 
['No Deposit' 'Refundable' 'Non Refund']

customer_type: 
['Transient' 'Contract' 'Transient-Party' 'Group']

year: 
[2015 2014 2016 2017]

month: 
[ 7  5  4  6  3  8  9  1 11 10 12  2]

day: 
[ 1  2  3  6 22 23  5  7  8 11 15 16 29 19 18  9 13  4 12 26 17 10 20 14
 30 28 25 21 27 24 31]



In [ ]:
# encoding categorical variables

cat_df['hotel'] = cat_df['hotel'].map({'Resort Hotel' : 0, 'City Hotel' : 1})

cat_df['meal'] = cat_df['meal'].map({'BB' : 0, 'FB': 1, 'HB': 2, 'SC': 3, 'Undefined': 4})

cat_df['market_segment'] = cat_df['market_segment'].map({'Direct': 0, 'Corporate': 1, 'Online TA': 2, 'Offline TA/TO': 3,
                                                           'Complementary': 4, 'Groups': 5, 'Undefined': 6, 'Aviation': 7})

cat_df['distribution_channel'] = cat_df['distribution_channel'].map({'Direct': 0, 'Corporate': 1, 'TA/TO': 2, 'Undefined': 3,
                                                                       'GDS': 4})

cat_df['reserved_room_type'] = cat_df['reserved_room_type'].map({'C': 0, 'A': 1, 'D': 2, 'E': 3, 'G': 4, 'F': 5, 'H': 6,
                                                                   'L': 7, 'B': 8 , 'P': 9})

cat_df['assigned_room_type'] = cat_df['assigned_room_type'].map({'C': 0, 'A': 1, 'D': 2, 'E': 3, 'G': 4, 'F': 5, 'H': 6,
                                                                   'L': 7, 'B': 8 , 'P': 9, 'I': 10 , 'K' : 11})

cat_df['deposit_type'] = cat_df['deposit_type'].map({'No Deposit': 0, 'Refundable': 1, 'Non Refund': 3})

cat_df['customer_type'] = cat_df['customer_type'].map({'Transient': 0, 'Contract': 1, 'Transient-Party': 2, 'Group': 3})

cat_df['year'] = cat_df['year'].map({2015: 0, 2014: 1, 2016: 2, 2017: 3})

In [ ]:
cat_df.head()

hotel  meal  market_segment  distribution_channel  reserved_room_type  \
0      0     0               0                     0                   0   
1      0     0               0                     0                   0   
2      0     0               0                     0                   1   
3      0     0               1                     1                   1   
4      0     0               2                     2                   1   

   assigned_room_type  deposit_type  customer_type  year  month  day  
0                   0             0              0     0      7    1  
1                   0             0              0     0      7    1  
2                   0             0              0     0      7    2  
3                   1             0              0     0      7    2  
4                   1             0              0     0      7    3

In [ ]:
num_df = df.drop(columns = cat_cols, axis = 1)
num_df['total_stays_in_week_nights'] = num_df['stays_in_weekend_nights'] + num_df['stays_in_week_nights']
num_df.drop(columns = ['stays_in_weekend_nights' , 'stays_in_week_nights'] , axis = 1)
num_df.head()

lead_time  arrival_date_week_number  arrival_date_day_of_month  \
0        342                        27                          1   
1        737                        27                          1   
2          7                        27                          1   
3         13                        27                          1   
4         14                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  is_repeated_guest  \
0                        0                     0                  0   
1                        0                     0                  0   
2                        0                     1                  0   
3                        0                     1                  0   
4                        0                     2                  0   

   previous_cancellations  previous_bookings_not_canceled  booking_changes  \
0                       0                               0                3   
1                       0                               0                4   
2                       0                               0                0   
3                       0                               0                0   
4                       0                               0                0   

   agent  company  days_in_waiting_list   adr  required_car_parking_spaces  \
0    0.0      0.0                     0   0.0                            0   
1    0.0      0.0                     0   0.0                            0   
2    0.0      0.0                     0  75.0                            0   
3  304.0      0.0                     0  75.0                            0   
4  240.0      0.0                     0  98.0                            0   

   total_of_special_requests  total_guests  total_stays_in_week_nights  
0                          0           2.0                           0  
1                          0           2.0                           0  
2                          0           1.0                           1  
3                          0           1.0                           1  
4                          1           2.0                           2

In [ ]:
num_df.var() #variance

lead_time                         11419.721511
arrival_date_week_number            185.099790
arrival_date_day_of_month            77.102966
stays_in_weekend_nights               0.997229
stays_in_week_nights                  3.641554
is_repeated_guest                     0.030894
previous_cancellations                0.712904
previous_bookings_not_canceled        2.242317
booking_changes                       0.425503
agent                             11479.398119
company                            2909.942594
days_in_waiting_list                309.574203
adr                                2553.866100
required_car_parking_spaces           0.060168
total_of_special_requests             0.628529
total_guests                          0.521853
total_stays_in_week_nights            6.540493
dtype: float64

In [ ]:
#normalizing variables

num_df['lead_time'] = np.log(num_df['lead_time'] + 1)
num_df['arrival_date_week_number'] = np.log(num_df['arrival_date_week_number'] + 1)
num_df['arrival_date_day_of_month'] = np.log(num_df['arrival_date_day_of_month'] + 1)
num_df['agent'] = np.log(num_df['agent'] + 1)
num_df['company'] = np.log(num_df['company'] + 1)
num_df['adr'] = np.log(num_df['adr'] + 1)
num_df['days_in_waiting_list'] = np.log(num_df['days_in_waiting_list'] + 1)


In [ ]:
num_df['adr'] = num_df['adr'].fillna(value = num_df['adr'].mean())

In [ ]:
num_df.head()

lead_time  arrival_date_week_number  arrival_date_day_of_month  \
0   5.837730                  3.332205                   0.693147   
1   6.603944                  3.332205                   0.693147   
2   2.079442                  3.332205                   0.693147   
3   2.639057                  3.332205                   0.693147   
4   2.708050                  3.332205                   0.693147   

   stays_in_weekend_nights  stays_in_week_nights  is_repeated_guest  \
0                        0                     0                  0   
1                        0                     0                  0   
2                        0                     1                  0   
3                        0                     1                  0   
4                        0                     2                  0   

   previous_cancellations  previous_bookings_not_canceled  booking_changes  \
0                       0                               0                3   
1                       0                               0                4   
2                       0                               0                0   
3                       0                               0                0   
4                       0                               0                0   

      agent  company  days_in_waiting_list       adr  \
0  0.000000      0.0                   0.0  0.000000   
1  0.000000      0.0                   0.0  0.000000   
2  0.000000      0.0                   0.0  4.330733   
3  5.720312      0.0                   0.0  4.330733   
4  5.484797      0.0                   0.0  4.595120   

   required_car_parking_spaces  total_of_special_requests  total_guests  \
0                            0                          0           2.0   
1                            0                          0           2.0   
2                            0                          0           1.0   
3                            0                          0           1.0   
4                            0                          1           2.0   

   total_stays_in_week_nights  
0                           0  
1                           0  
2                           1  
3                           1  
4                           2

**Model Prep**

In [ ]:
features = pd.concat([cat_df, num_df], axis = 1)
X = features.drop(columns = ['adr'])
y = features['adr']

In [ ]:
features.head()

hotel  meal  market_segment  distribution_channel  reserved_room_type  \
0      0     0               0                     0                   0   
1      0     0               0                     0                   0   
2      0     0               0                     0                   1   
3      0     0               1                     1                   1   
4      0     0               2                     2                   1   

   assigned_room_type  deposit_type  customer_type  year  month  ...  \
0                   0             0              0     0      7  ...   
1                   0             0              0     0      7  ...   
2                   0             0              0     0      7  ...   
3                   1             0              0     0      7  ...   
4                   1             0              0     0      7  ...   

   previous_bookings_not_canceled  booking_changes     agent  company  \
0                               0                3  0.000000      0.0   
1                               0                4  0.000000      0.0   
2                               0                0  0.000000      0.0   
3                               0                0  5.720312      0.0   
4                               0                0  5.484797      0.0   

   days_in_waiting_list       adr  required_car_parking_spaces  \
0                   0.0  0.000000                            0   
1                   0.0  0.000000                            0   
2                   0.0  4.330733                            0   
3                   0.0  4.330733                            0   
4                   0.0  4.595120                            0   

   total_of_special_requests  total_guests  total_stays_in_week_nights  
0                          0           2.0                           0  
1                          0           2.0                           0  
2                          0           1.0                           1  
3                          0           1.0                           1  
4                          1           2.0                           2  

[5 rows x 28 columns]

In [ ]:
X.shape, y.shape

((119390, 27), (119390,))

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20 , random_state= 42)

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

**Model Building**

Linear Regression

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred_lr = lr.predict(X_test)

lr_r2 = r2_score(y_test, y_pred_lr)
mse = mean_squared_error(y_test, y_pred_lr)
mae = mean_absolute_error(y_test, y_pred_lr)

print(f"R2 Score of Linear Regression is : {lr_r2}")
print(f"MSE of Linear Regression is : {mse}")
print(f"MAE of Linear Regression is : {mae}")


R2 Score of Linear Regression is : 0.25561729766249675
MSE of Linear Regression is : 0.37500599126226114
MAE of Linear Regression is : 0.35040526907567754


Decision Tree Regression

In [ ]:
from sklearn.tree import DecisionTreeRegressor
dtr = DecisionTreeRegressor()
dtr.fit(X_train, y_train)

y_pred_dtr = dtr.predict(X_test)

dtr_r2 = r2_score(y_test, y_pred_dtr)
mse = mean_squared_error(y_test, y_pred_dtr)
mae = mean_absolute_error(y_test, y_pred_dtr)

print(f"R2 Score of Decision Tree Regression is : {dtr_r2}")
print(f"MSE of Decision Tree Regression is : {mse}")
print(f"MAE of Decision Tree Regression is : {mae}")

R2 Score of Decision Tree Regression is : 0.6944506608455658
MSE of Decision Tree Regression is : 0.15393000461902928
MAE of Decision Tree Regression is : 0.10720851357636124


K-NN Regression

In [ ]:
from sklearn.neighbors import KNeighborsRegressor
knn = KNeighborsRegressor()
knn.fit(X_train, y_train)

y_pred_knn = knn.predict(X_test)

knn_r2 = r2_score(y_test, y_pred_knn)
mse = mean_squared_error(y_test, y_pred_knn)
mae = mean_absolute_error(y_test, y_pred_knn)

print(f"R2 Score of K-NN Regression is : {knn_r2}")
print(f"MSE of K-NN Regression is : {mse}")
print(f"MAE of K-NN Regression is : {mae}")


R2 Score of K-NN Regression is : 0.5745017203088535
MSE of K-NN Regression is : 0.21435802263392562
MAE of K-NN Regression is : 0.1954455932096827


Random Forest Regression

In [ ]:
from sklearn.ensemble import RandomForestRegressor

rfr = RandomForestRegressor()
rfr.fit(X_train, y_train)

y_pred_rfr = rfr.predict(X_test)

rfr_r2 = r2_score(y_test, y_pred_rfr)
mse = mean_squared_error(y_test, y_pred_rfr)
mae = mean_absolute_error(y_test, y_pred_rfr)

print(f"R2 Score of Random Forest Regression is : {rfr_r2}")
print(f"MSE of Random Forest Regression is : {mse}")
print(f"MAE of Random Forest Regression is : {mae}")


R2 Score of Random Forest Regression is : 0.8453946766249075
MSE of Random Forest Regression is : 0.07788725122794672
MAE of Random Forest Regression is : 0.08938153002384269


Gradient Boosting Regression

In [ ]:
from sklearn.ensemble import GradientBoostingRegressor

gbr = GradientBoostingRegressor()
gbr.fit(X_train, y_train)

y_pred_gbr = gbr.predict(X_test)

gbr_r2 = r2_score(y_test, y_pred_gbr)
mse = mean_squared_error(y_test, y_pred_gbr)
mae = mean_absolute_error(y_test, y_pred_gbr)

print(f"R2 Score of Gradient Boosting Regression is : {gbr_r2}")
print(f"MSE of Gradient Boosting Regression is : {mse}")
print(f"MAE of Gradient Boosting Regression is : {mae}")

R2 Score of Gradient Boosting Regression is : 0.731207073216335
MSE of Gradient Boosting Regression is : 0.13541281606392083
MAE of Gradient Boosting Regression is : 0.2005288933234459


XGBoost Regression

In [ ]:
from xgboost import XGBRegressor

xgb = XGBRegressor()
xgb.fit(X_train, y_train)

y_pred_xgb = xgb.predict(X_test)

xgb_r2 = r2_score(y_test, y_pred_xgb)
mse = mean_squared_error(y_test, y_pred_xgb)
mae = mean_absolute_error(y_test, y_pred_xgb)

print(f"R2 Score of XGBoost Regression is : {xgb_r2}")
print(f"MSE of XGBoost Regression is : {mse}")
print(f"MAE of XGBoost Regression is : {mae}")

R2 Score of XGBoost Regression is : 0.8439835362360331
MSE of XGBoost Regression is : 0.07859815718892398
MAE of XGBoost Regression is : 0.12375081693354803


LightGBM


In [ ]:
from lightgbm import LGBMRegressor

lgbm = LGBMRegressor()
lgbm.fit(X_train, y_train)

y_pred_lgbm = lgbm.predict(X_test)

lgbm_r2 = r2_score(y_test, y_pred_lgbm)
mse = mean_squared_error(y_test, y_pred_lgbm)
mae = mean_absolute_error(y_test, y_pred_lgbm)

print(f"R2 Score of LightGBM Regression is : {lgbm_r2}")
print(f"MSE of LightGBM Regression is : {mse}")
print(f"MAE of LightGBM Regression is : {mae}")

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.020757 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1140
[LightGBM] [Info] Number of data points in the train set: 95512, number of used features: 27
[LightGBM] [Info] Start training from score 4.475267
R2 Score of LightGBM Regression is : 0.8284260694138578
MSE of LightGBM Regression is : 0.0864357160801492
MAE of LightGBM Regression is : 0.14072841633614505


CatBoost

In [ ]:
!pip install catboost


In [ ]:
from catboost import CatBoostRegressor

catboost = CatBoostRegressor(silent=True)
catboost.fit(X_train, y_train)

y_pred_catboost = catboost.predict(X_test)

catboost_r2 = r2_score(y_test, y_pred_catboost)
mse = mean_squared_error(y_test, y_pred_catboost)
mae = mean_absolute_error(y_test, y_pred_catboost)

print(f"R2 Score of CatBoost Regression is : {catboost_r2}")
print(f"MSE of CatBoost Regression is : {mse}")
print(f"MAE of CatBoost Regression is : {mae}")



R2 Score of CatBoost Regression is : 0.8443550688009468
MSE of CatBoost Regression is : 0.07841098607740542
MAE of CatBoost Regression is : 0.12482439341752002


Support Vector Regression (SVR)

In [ ]:
from sklearn.svm import SVR

svr = SVR()
svr.fit(X_train, y_train)

y_pred_svr = svr.predict(X_test)


In [ ]:
svr_r2 = r2_score(y_test, y_pred_svr)
mse = mean_squared_error(y_test, y_pred_svr)
mae = mean_absolute_error(y_test, y_pred_svr)

print(f"R2 Score of Support Vector Regression is : {svr_r2}")
print(f"MSE of Support Vector Regression is : {mse}")
print(f"MAE of Support Vector Regression is : {mae}")

Voting Regression

In [ ]:
from sklearn.ensemble import VotingRegressor

ensemble = VotingRegressor(estimators=[
    ('lr', lr), ('dtr', dtr), ('knn', knn),
    ('rfr', rfr), ('gbr', gbr), ('xgb', xgb),
    ('lgbm', lgbm), ('catboost', catboost)
])

ensemble.fit(X_train, y_train)

y_pred_ensemble = ensemble.predict(X_test)

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.021660 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1140
[LightGBM] [Info] Number of data points in the train set: 95512, number of used features: 27
[LightGBM] [Info] Start training from score 4.475267


In [ ]:
ensemble_r2 = r2_score(y_test, y_pred_ensemble)
mse = mean_squared_error(y_test, y_pred_ensemble)
mae = mean_absolute_error(y_test, y_pred_ensemble)

print(f"R2 Score of Voting Regressor is: {ensemble_r2}")
print(f"MSE of Voting Regressor is: {mse}")
print(f"MAE of Voting Regressor is: {mae}")

R2 Score of Voting Regressor is: 0.8277135912199458
MSE of Voting Regressor is: 0.08679464918071926
MAE of Voting Regressor is: 0.13712895022052088


Multilayer Perceptron (MLP)


In [ ]:
from sklearn.neural_network import MLPRegressor

mlp = MLPRegressor(hidden_layer_sizes=(100, 50), activation='relu', solver='adam', random_state=42 , max_iter=300)

mlp.fit(X_train, y_train)

y_pred_mlp = mlp.predict(X_test)

mlp_r2 = r2_score(y_test, y_pred_mlp)
mse = mean_squared_error(y_test, y_pred_mlp)
mae = mean_absolute_error(y_test, y_pred_mlp)

print(f"R2 Score of MLP Regressor is: {mlp_r2}")
print(f"MSE of MLP Regressor is: {mse}")
print(f"MAE of MLP Regressor is: {mae}")

R2 Score of MLP Regressor is: 0.7738600461306764
MSE of MLP Regressor is: 0.11392505131898893
MAE of MLP Regressor is: 0.16373924512364946


ANN

In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Convert data to PyTorch tensors
X_train_tensor = torch.tensor(X_train.values, dtype=torch.float32).to(device)
y_train_tensor = torch.tensor(y_train.values, dtype=torch.float32).reshape(-1, 1).to(device)
X_test_tensor = torch.tensor(X_test.values, dtype=torch.float32).to(device)
y_test_tensor = torch.tensor(y_test.values, dtype=torch.float32).reshape(-1, 1).to(device)

train_dataset = TensorDataset(X_train_tensor, y_train_tensor)
train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)

class ANN(nn.Module):
    def __init__(self, input_size):
        super(ANN, self).__init__()
        self.fc1 = nn.Linear(input_size, 128)
        self.fc2 = nn.Linear(128, 64)
        self.fc3 = nn.Linear(64, 1)
        self.relu = nn.ReLU()

    def forward(self, x):
        x = self.relu(self.fc1(x))
        x = self.relu(self.fc2(x))
        x = self.fc3(x)
        return x

input_size = X_train_tensor.shape[1]
model = ANN(input_size).to(device)
criterion = nn.MSELoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Train
num_epochs = 100
for epoch in range(num_epochs):
    running_loss = 0.0
    for inputs, targets in train_loader:
        inputs, targets = inputs.to(device), targets.to(device)

        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, targets)
        loss.backward()
        optimizer.step()

        running_loss += loss.item()

    print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {running_loss/len(train_loader):.4f}')


In [ ]:
# Evaluate
model.eval()
with torch.no_grad():
    y_pred_ann = model(X_test_tensor)
    ann_r2 = r2_score(y_test_tensor.cpu().numpy(), y_pred_ann.cpu().numpy())
    mse = mean_squared_error(y_test_tensor.cpu().numpy(), y_pred_ann.cpu().numpy())
    mae = mean_absolute_error(y_test_tensor.cpu().numpy(), y_pred_ann.cpu().numpy())

print(f"R2 Score of ANN Regressor is: {ann_r2}")
print(f"MSE of ANN Regressor is: {mse}")
print(f"MAE of ANN Regressor is: {mae}")


#Score:  ANN	0.769392


In [ ]:
models = pd.DataFrame({
                        'Model': ['Linear Regression', 'Decision Tree Regression', 'K-NN Regression', 'Random Forest Regression',
                                 'Gradient Boosting Regression', 'XGBoost Regression', 'LightGBM',
                                 'CatBoost','Voting Regression', 'Multilayer Perceptron' , 'ANN'],
                        'Score' : [lr_r2 , dtr_r2 , knn_r2 , rfr_r2, gbr_r2 , xgb_r2 , lgbm_r2, catboost_r2, ensemble_r2 , mlp_r2 , 0.769392 ]  })

models.sort_values(by = 'Score', ascending = False)

Model     Score
3       Random Forest Regression  0.845395
7                       CatBoost  0.844355
5             XGBoost Regression  0.843984
6                       LightGBM  0.828426
8              Voting Regression  0.827714
9          Multilayer Perceptron  0.773860
10                           ANN  0.769392
4   Gradient Boosting Regression  0.731207
1       Decision Tree Regression  0.694451
2                K-NN Regression  0.574502
0              Linear Regression  0.255617

I changed my test_size to 0.20 and used StandardScaler()
in addition to hyperparam tuning to increase the accuracy even further

**Tuning and checking models for overfitting**

Tuning Random Forest Further

In [ ]:
from sklearn.ensemble import RandomForestRegressor

rfr = RandomForestRegressor(n_estimators = 300)
rfr.fit(X_train, y_train)

y_train_pred = rfr.predict(X_train)
y_test_pred = rfr.predict(X_test)

train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

print("R2 score on training set:", train_r2)
print("R2 score on test set:", test_r2)

#overfitting

R2 score on training set: 0.9628990091019416
R2 score on test set: 0.8459533603721444


Tuning XGBoost Regression further

In [ ]:
from xgboost import XGBRegressor
from sklearn.model_selection import GridSearchCV

xgb = XGBRegressor(colsample_bytree= 0.9, learning_rate= 0.1, max_depth= 5,
                   n_estimators= 300, subsample= 0.9,random_state=42)

xgb.fit(X_train, y_train)

y_train_pred = xgb.predict(X_train)
y_test_pred = xgb.predict(X_test)

train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

print("R2 score on training set:", train_r2)
print("R2 score on test set:", test_r2)

R2 score on training set: 0.8424488130107468
R2 score on test set: 0.8362141910535276


Tuning CatBoost Further


In [ ]:
!pip install catboost


In [ ]:
import matplotlib.pyplot as plt
from catboost import CatBoostRegressor, Pool
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

catboost = CatBoostRegressor(
    iterations=1000,
    learning_rate=0.1,
    depth=7,
    l2_leaf_reg=7,
    silent=True,
    random_state=42
)

catboost.fit(X_train, y_train)

y_train_pred = catboost.predict(X_train)
y_test_pred = catboost.predict(X_test)

train_r2 = r2_score(y_train, y_train_pred)
test_r2 = r2_score(y_test, y_test_pred)

print("R2 score on training set:", train_r2)
print("R2 score on test set:", test_r2)


R2 score on training set: 0.8692440078860733
R2 score on test set: 0.8504236786182734


I've achived **0.8504236786182734** accuracy using **CatBoost** Algorithm
with a slight overfit and **0.8362141910535276** score with **XGBoost** with minimum overfitting.

there is room to improve using feature engineering and adding more feature data.



**TODO** : Improve the model using more feature engineering and hyper-parameter tunings.